File: backend/app/services/movements.py
"""The movement log of the warehouse: which compartment a device left, which it entered,
when, and how long it had been there.

The business owner's second instruction of 24.09.2026: "from which storage compartment or
part of the warehouse is the piece being moved, so we understand the movement and the
turnaround." Three reads in this system said they could not answer it: ``warehouse.py``
derives its throughput from stock and dwell, ``capacity_plan.py`` inverts a dwell that is
the age of the stock still present, and ``tco_device.py`` cannot cost the warehouse days
of a finished life. All three say a measured figure needs the movement log.

The log existed in substance. ``asset_service.transition`` has written an ``AssetEvent``
with ``from_status -> to_status`` since the lifecycle log was added, and a compartment is
defined by its statuses, so every transition already said which compartment a device
left and entered. Two things were missing, and both are on the event row now (migration
f7a9b1c3d5e4): the day of the move on the fleet's own calendar (``effective_date``) and
the stay it ended (``from_since``, ``dwell_days``), written by the asset service at the
moment of the move. A third gap is not in the schema but in the data: the seed streams
the fleet in bulk and writes no events, so the 431,200 seeded serials carry no history,
and every figure here is measured from the moves made since, by the simulation tab, the
console's transition button or an integration. The reads say so rather than pad it.

**Measured, not derived.** A flow here is a count of moves in a window over the days of
the window. A dwell here is the stay of a device that actually left, from the two stamps
on the row that moved it. Both are what the other modules derive by Little's law from the
stock still present, and where a compartment shows both, they stand side by side.

**Everything aggregates in the database.** The window read is one grouped query over
(from status, to status, dwell days) on an index that carries all three, plus a distinct
count of the devices that moved; a device's path is the handful of rows behind one serial.

**No fake zeros.** A window with no dated move says so; a stay whose start the seed did
not record is counted as a move of unknown length, never as zero days.
"""
from __future__ import annotations

from collections import defaultdict
from datetime import date, timedelta
from typing import Optional

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.models.catalog import Product
from app.models.flow import WAREHOUSE_STATUSES, Asset, AssetEvent, AssetEventType, AssetStatus, Location
from app.services import warehouse
from app.services.exceptions import NotFoundError

DEFAULT_DAYS = 30
MAX_DAYS = 365

# Where a device is when it is not in a compartment, as the operator names it.
OUTSIDE_LABEL = {
    AssetStatus.RENTED: "At customers", AssetStatus.SOLD: "Sold", AssetStatus.RECYCLED: "Recycled",
    AssetStatus.DEPLOYED: "Deployed", AssetStatus.MAINTENANCE: "Maintenance",
    AssetStatus.DECOMMISSIONED: "Decommissioned", AssetStatus.DISPOSED: "Disposed",
}
BORN_LABEL = "On order"     # a receipt: the device had no compartment before, it was on an order line

FLOW_BASIS = "measured: moves out of the compartment in the window, over the days the log covers inside it, times seven"
DWELL_BASIS = ("measured: the stay of every device that left the compartment in the window, from the day it entered "
               "(the dwell start it carried) to the day it left; a device whose entry the seed did not record is a "
               "move of unknown length and is left out of the days")
DERIVED_BASIS = warehouse.THROUGHPUT_BASIS
HISTORY_NOTE = ("the seeded fleet carries no movement history: the seed writes devices in bulk without events. Every "
                "move made since, on the Simulation tab, through the console's transition button or by an integration, "
                "is logged with its day and the stay it ended, and that is what is measured here")


def _as_date(v) -> Optional[date]:
    if v is None:
        return None
    return v if isinstance(v, date) else date.fromisoformat(str(v)[:10])


def _status(v) -> Optional[AssetStatus]:
    if v is None:
        return None
    return v if isinstance(v, AssetStatus) else AssetStatus(str(v))


def label(status: Optional[AssetStatus]) -> str:
    """A status as a place: the compartment it belongs to, or where the device is instead."""
    if status is None:
        return BORN_LABEL
    comp = warehouse.STATION_OF_STATUS.get(status)
    if comp is not None:
        return comp.name
    return OUTSIDE_LABEL.get(status, status.value)


def code(status: Optional[AssetStatus]) -> Optional[str]:
    """The compartment code of a status, or None outside the warehouse."""
    comp = warehouse.STATION_OF_STATUS.get(status) if status is not None else None
    return comp.code if comp else None


def _quantiles(hist: dict[int, int]) -> dict:
    """Median, p90, mean and the longest stay from {days: how many}; None when nothing is dated."""
    dated = sum(hist.values())
    if not dated:
        return {"median_days": None, "p90_days": None, "mean_days": None, "max_days": None}
    return {
        "median_days": warehouse._quantile_from_histogram(hist, 0.5),
        "p90_days": warehouse._quantile_from_histogram(hist, 0.9),
        "mean_days": round(sum(d * n for d, n in hist.items()) / dated, 1),
        "max_days": max(hist),
    }


def window(db: Session, *, days: int = DEFAULT_DAYS, today: Optional[date] = None) -> dict:
    """The moves of the last ``days`` days, by pair of compartments and rolled up per compartment.

    One grouped read on the window index: (from status, to status, days in the from
    compartment) with a count, a few hundred rows however many moves the log holds. A
    pure location move (no status change) is not a compartment movement and is left out;
    a receipt is the move from an order line into new stock.
    """
    today = today or date.today()
    days = max(1, min(int(days), MAX_DAYS))
    since = today - timedelta(days=days)
    in_window = (AssetEvent.effective_date > since, AssetEvent.effective_date <= today)

    rows = db.execute(
        select(AssetEvent.from_status, AssetEvent.to_status, AssetEvent.dwell_days, func.count())
        .where(*in_window, AssetEvent.to_status.is_not(None))
        .group_by(AssetEvent.from_status, AssetEvent.to_status, AssetEvent.dwell_days)
    ).all()
    devices = int(db.scalar(select(func.count(func.distinct(AssetEvent.asset_id)))
                            .where(*in_window, AssetEvent.to_status.is_not(None))) or 0)
    undated = int(db.scalar(select(func.count()).select_from(AssetEvent)
                            .where(AssetEvent.effective_date.is_(None), AssetEvent.to_status.is_not(None))) or 0)
    first, last = db.execute(select(func.min(AssetEvent.effective_date), func.max(AssetEvent.effective_date))
                             .where(AssetEvent.to_status.is_not(None))).one()
    first, last = _as_date(first), _as_date(last)
    # A flow per week is moves over days, and the days are the ones the log covers: from the first
    # logged move inside the window to today. A seeded fleet has no history, so a 30-day window that
    # holds two days of moves would otherwise read as a fifteenth of the real flow.
    first_in = _as_date(db.scalar(select(func.min(AssetEvent.effective_date)).where(*in_window, AssetEvent.to_status.is_not(None))))
    covered = ((today - first_in).days + 1) if first_in else 0

    def per_week(n: int) -> float:
        return round(n / covered * 7, 1) if covered else 0.0

    pairs: dict[tuple, dict] = {}
    for frm, to, dwell, n in rows:
        frm, to, n = _status(frm), _status(to), int(n)
        p = pairs.setdefault((frm, to), {"hist": {}, "units": 0, "unknown_units": 0})
        p["units"] += n
        if dwell is None:
            p["unknown_units"] += n
        else:
            p["hist"][int(dwell)] = p["hist"].get(int(dwell), 0) + n

    order = {c.code: i for i, c in enumerate(warehouse.COMPARTMENTS, start=1)}

    def _rank(st: Optional[AssetStatus]) -> int:
        c = code(st)
        return order.get(c, 0) if c else (0 if st is None else 99)

    pair_rows = []
    for (frm, to), p in sorted(pairs.items(), key=lambda kv: (_rank(kv[0][0]), _rank(kv[0][1]), -kv[1]["units"])):
        dated = sum(p["hist"].values())
        pair_rows.append({
            "from_status": (frm.value if frm else None), "to_status": to.value,
            "from_code": code(frm), "to_code": code(to), "from_name": label(frm), "to_name": label(to),
            "units": p["units"], "dated_units": dated, "unknown_units": p["unknown_units"],
            "per_week": per_week(p["units"]),
            **_quantiles(p["hist"]),
            "dwell_reason": (None if dated else ("a receipt ends no stay: the device was on an order line" if frm is None
                                                 else "no stay measured: the devices that moved carried no dwell start")),
        })

    # per compartment: what came in, what went out, the measured flow and the measured finished stay,
    # next to what the Warehouse tab derives from the stock still present
    derived = {c["code"]: c for c in warehouse.compartments(db, today=today)["compartments"]}
    comp_rows = []
    for step, c in enumerate(warehouse.COMPARTMENTS, start=1):
        sts = set(c.statuses)
        into = sum(p["units"] for (f, t), p in pairs.items() if t in sts and f not in sts)
        out = sum(p["units"] for (f, t), p in pairs.items() if f in sts and t not in sts)
        hist: dict[int, int] = defaultdict(int)
        unknown = 0
        for (f, t), p in pairs.items():
            if f in sts and t not in sts:
                unknown += p["unknown_units"]
                for d, n in p["hist"].items():
                    hist[d] += n
        q = _quantiles(dict(hist))
        d = derived.get(c.code, {})
        comp_rows.append({
            "code": c.code, "name": c.name, "step": step, "stage": c.stage,
            "units_in": into, "units_out": out, "in_per_week": per_week(into), "out_per_week": per_week(out),
            "flow_basis": FLOW_BASIS,
            "dated_out": sum(hist.values()), "unknown_out": unknown, **q, "dwell_basis": DWELL_BASIS,
            "dwell_reason": (None if hist else ("nothing left this compartment in the window" if not out
                                                else "the devices that left carried no dwell start")),
            "target_dwell_days": c.target_dwell_days,
            "on_hand": d.get("on_hand"),
            "derived_units_per_week": d.get("units_per_week"), "derived_mean_days": d.get("mean_days"),
            "derived_basis": DERIVED_BASIS,
        })

    moves = sum(p["units"] for p in pairs.values())
    return {
        "as_of": today, "days": days, "since": since, "covered_days": covered,
        "coverage_basis": ("a flow per week is the moves over the days the log covers inside the window, from the first logged move to "
                           "today, times seven; a window the log does not fill is not padded with empty days"),
        "moves": moves, "devices": devices, "pairs_count": len(pair_rows),
        "undated_events": undated,
        "undated_reason": (None if not undated else f"{undated:,} logged moves carry no day: written before the movement log had one"),
        "first_move": first, "last_move": last,
        "history_note": HISTORY_NOTE,
        "reason": (None if moves else (f"no move logged in the last {days} days" if last else "no move logged yet")),
        "pairs": pair_rows,
        "compartments": comp_rows,
    }


def path(db: Session, serial: str, *, today: Optional[date] = None) -> dict:
    """One device's way through the chain: every logged move with its day and the stay it ended,
    then the compartment it is in now and how long it has been there so far."""
    today = today or date.today()
    row = db.execute(
        select(Asset, Product.name, Product.category, Location.code)
        .join(Product, Product.id == Asset.product_id)
        .outerjoin(Location, Location.id == Asset.current_location_id)
        .where(Asset.serial_number == serial)
    ).first()
    if row is None:
        raise NotFoundError(f"No device with serial {serial!r}")
    asset, product, family, station = row
    # Append-only, so the write order is the order of the moves; the day is on the row for the reading.
    events = db.execute(select(AssetEvent).where(AssetEvent.asset_id == asset.id)
                        .order_by(AssetEvent.date_created, AssetEvent.id)).scalars().all()
    steps = []
    warehouse_days = 0
    customer_days = 0
    unknown = 0
    for e in events:
        if e.event_type == AssetEventType.MOVED:
            steps.append({"kind": "moved", "from_status": None, "to_status": None, "from_name": None, "to_name": None,
                          "from_code": None, "to_code": None, "effective_date": _as_date(e.effective_date),
                          "from_since": None, "dwell_days": None, "actor": e.actor, "note": e.note, "logged_at": e.date_created})
            continue
        frm, to = _status(e.from_status), _status(e.to_status)
        if e.dwell_days is None and frm is not None:
            unknown += 1
        elif e.dwell_days is not None and frm in WAREHOUSE_STATUSES:
            warehouse_days += int(e.dwell_days)
        elif e.dwell_days is not None and frm == AssetStatus.RENTED:
            customer_days += int(e.dwell_days)
        steps.append({
            "kind": ("received" if e.event_type == AssetEventType.RECEIVED else "moved_status"),
            "from_status": (frm.value if frm else None), "to_status": (to.value if to else None),
            "from_name": (label(frm) if (frm or e.event_type == AssetEventType.RECEIVED) else None),
            "to_name": (label(to) if to else None),
            "from_code": code(frm), "to_code": code(to),
            "effective_date": _as_date(e.effective_date), "from_since": _as_date(e.from_since), "dwell_days": e.dwell_days,
            "actor": e.actor, "note": e.note, "logged_at": e.date_created,
        })
    since = _as_date(asset.status_since)
    return {
        "serial_number": asset.serial_number, "asset_id": asset.id, "product": product, "family": family,
        "cycle_no": int(asset.cycle_no or 0), "grade": asset.grade,
        "status": asset.status.value, "station_name": label(asset.status), "station_code": code(asset.status),
        "location_code": station, "since": since, "days_so_far": ((today - since).days if since else None),
        "steps": steps, "moves_logged": len(steps),
        "warehouse_days_measured": warehouse_days, "customer_days_measured": customer_days, "unknown_stays": unknown,
        "history_reason": (None if steps else "no move logged for this device: " + HISTORY_NOTE),
        "as_of": today,
    }

File: frontend/movements.js
"use strict";
/* ============================================================================
   SCM Master — Movements (device-as-a-service). Loads after ordering.js.

   The owner's instruction: "from which storage compartment or part of the
   warehouse is the piece being moved, so we understand the movement and the
   turnaround." The movement log: every move between compartments in a window,
   by pair, with how many and how long they had been in the compartment they
   left, measured from the moves themselves; per compartment the measured flow
   and the measured finished stay next to what the Warehouse tab derives from
   the stock still present; and one serial opened to its own path through the
   chain with the days in each station.

   Every figure comes from /movements and is measured from logged moves. The
   seeded fleet carries no history, and the screen says so: the log fills from
   the Simulation tab, the console's transition button and integrations.
============================================================================ */

ICONS.route = '<circle cx="6" cy="18" r="2.5"/><circle cx="18" cy="6" r="2.5"/><path d="M8.5 18h5a3 3 0 0 0 0-6h-3a3 3 0 0 1 0-6h5"/>';
CRUMBS.movements = "Movements";

let MV = { days: 30, view: null, serial: "", path: null };
const MV_WINDOWS = [7, 30, 90, 365];

const mvDays = (v) => v == null ? `<span class="wh-na">n/a</span>` : `${num(Math.round(v))} d`;
const mvMeasured = (title) => `<span class="wh-derived om-measured" title="${esc(title)}">measured</span>`;

function mvPairsHtml(V) {
  if (!V.pairs.length) return `<div class="muted" style="font-size:12.5px">${esc(V.reason || "No move in this window.")}</div>`;
  const rows = V.pairs.map((p) => `<tr>
      <td><div class="mv-pair"><span class="mv-pair__from">${esc(p.from_name)}</span>${icon("arrow", 12)}<span class="mv-pair__to">${esc(p.to_name)}</span></div>
        <div class="wh-note">${p.from_code ? esc(p.from_code) : "outside"} → ${p.to_code ? esc(p.to_code) : "outside"}</div></td>
      <td class="num" style="font-weight:600">${num(p.units)}</td>
      <td class="num">${num(p.per_week)}</td>
      <td class="num">${p.median_days == null ? `<span class="wh-na">n/a</span><div class="wh-note">${esc(p.dwell_reason || "")}</div>` : `<b>${mvDays(p.median_days)}</b><div class="wh-note">mean ${num(p.mean_days)} · p90 ${num(p.p90_days)} · longest ${num(p.max_days)}${p.unknown_units ? ` · ${num(p.unknown_units)} unknown` : ""}</div>`}</td>
    </tr>`).join("");
  return `<table class="tbl mv-pairs"><thead><tr><th>From → to</th><th class="num">Units</th><th class="num">A week</th><th class="num">Days in the compartment they left</th></tr></thead><tbody>${rows}</tbody></table>`;
}

function mvCompartmentsHtml(V) {
  const rows = V.compartments.map((c) => {
    const flow = c.units_out ? `<b>${num(c.out_per_week)}</b> / week ${mvMeasured(c.flow_basis)}` : `<span class="wh-na">none</span>`;
    const derivedFlow = c.derived_units_per_week == null ? `<span class="wh-na">n/a</span>` : `${num(Math.round(c.derived_units_per_week))} / week <span class="wh-derived" title="${esc(c.derived_basis)}">derived</span>`;
    const stay = c.median_days == null ? `<span class="wh-na">n/a</span><div class="wh-note">${esc(c.dwell_reason || "")}</div>`
      : `<b>${mvDays(c.median_days)}</b> median <div class="wh-note">mean ${num(c.mean_days)} d · p90 ${num(c.p90_days)} d · ${num(c.dated_out)} stays${c.unknown_out ? `, ${num(c.unknown_out)} unknown` : ""} · target ${c.target_dwell_days} d</div>`;
    const derivedStay = c.derived_mean_days == null ? `<span class="wh-na">n/a</span>` : `${num(Math.round(c.derived_mean_days))} d mean age <span class="wh-derived" title="${esc(c.derived_basis)}">derived</span>`;
    return `<tr>
      <td><div class="cell-prod__name">${c.step}. ${esc(c.name)}</div><div class="wh-note">${num(c.on_hand)} on hand</div></td>
      <td class="num">${num(c.units_in)}</td>
      <td class="num">${num(c.units_out)}</td>
      <td class="num">${flow}<div class="wh-note">${derivedFlow}</div></td>
      <td>${stay}<div class="wh-note">${derivedStay}</div></td>
    </tr>`;
  }).join("");
  return `<table class="tbl mv-comp"><thead><tr><th>Compartment</th><th class="num">In</th><th class="num">Out</th><th class="num">Flow · measured vs derived</th><th>Finished stay · measured vs derived</th></tr></thead><tbody>${rows}</tbody></table>`;
}

function mvPathHtml(p) {
  const head = `<div class="mv-path__head">
      <div><div class="sim-effect__title"><span class="ref">${esc(p.serial_number)}</span> · ${esc(p.product)}</div>
        <div class="wh-note">${esc(p.family || "")} · ${p.cycle_no === 0 ? "never rented" : `${p.cycle_no} rental${p.cycle_no === 1 ? "" : "s"}`}${p.grade ? ` · grade ${esc(p.grade)}` : ""}</div></div>
      <div class="mv-path__now">${statusPill(p.status)}<div class="wh-note">in <b>${esc(p.station_name)}</b>${p.since ? ` since ${fmtDate(p.since)}, <b>${num(p.days_so_far)} day${p.days_so_far === 1 ? "" : "s"}</b> so far` : ""}</div></div>
    </div>`;
  if (!p.steps.length) return head + `<div class="muted" style="font-size:12.5px;margin-top:10px">${esc(p.history_reason || "No move logged.")}</div>`;
  const steps = p.steps.map((s) => {
    const title = s.kind === "moved" ? "Moved within the warehouse" : `${esc(s.from_name || "On order")} ${icon("arrow", 11)} ${esc(s.to_name)}`;
    const stay = s.kind === "moved" ? "" : s.dwell_days == null ? (s.from_status ? `<span class="wh-na">stay unknown</span>` : "")
      : `<b>${num(s.dwell_days)} days</b> in ${esc(s.from_name)}${s.from_since ? ` (since ${fmtDate(s.from_since)})` : ""}`;
    return `<div class="log__entry"><div class="log__rail"><div class="log__dot" style="background:${s.to_code ? "var(--ts-brand-gold)" : "var(--ts-info)"}"></div><div class="log__line"></div></div>
      <div class="log__body"><div class="log__time">${s.effective_date ? fmtDate(s.effective_date) : "no day logged"}${s.actor ? " · " + esc(s.actor) : ""}</div>
        <div class="attn__title">${title}</div><div class="log__note">${stay}${s.note ? `${stay ? " · " : ""}${esc(s.note)}` : ""}</div></div></div>`;
  }).join("");
  return head + `<div class="log" style="margin-top:12px">${steps}</div>
    <div class="wh-note" style="margin-top:8px"><b>${num(p.warehouse_days_measured)} days</b> in the warehouse across the finished stays logged${p.customer_days_measured ? ` · ${num(p.customer_days_measured)} days at customers` : ""}${p.unknown_stays ? ` · ${num(p.unknown_stays)} stay${p.unknown_stays === 1 ? "" : "s"} of unknown length (entered before the log)` : ""} ${mvMeasured("the sum of the stays on the logged moves; the current stay is still open and not in this figure")}</div>`;
}

async function mvOpenSerial() {
  const host = $("#mv-path");
  const serial = ($("#mv-serial").value || "").trim();
  if (!host || !serial) return;
  MV.serial = serial;
  host.innerHTML = `<div class="muted" style="font-size:12.5px">Reading…</div>`;
  try {
    MV.path = await api(`/movements/serials/${encodeURIComponent(serial)}`);
    host.innerHTML = mvPathHtml(MV.path);
  } catch (e) {
    host.innerHTML = `<div class="muted" style="font-size:12.5px">${esc((e && e.message) || "Could not read")}</div>`;
  }
}

async function mvDraw(screen) {
  const V = await api(`/movements?days=${MV.days}`);
  MV.view = V;
  const stat = (label, ic, val, hint, hintCls = "", valCls = "") =>
    `<div class="stat"><div class="stat__label">${icon(ic, 14)} ${label}</div><div class="stat__val ${valCls}">${val}</div>${hint ? `<div class="stat__hint ${hintCls}">${hint}</div>` : ""}</div>`;
  const busiest = V.compartments.slice().sort((a, b) => b.units_out - a.units_out)[0];
  screen.innerHTML = `
    ${pageHead("Warehouse", "Movements", "Which compartment a device left, which it entered, when, and how long it had been there. Measured from the moves themselves, never from the stock still present: the movement log the Warehouse tab, the capacity plan and the device TCO said they needed. Open a serial to see its own way through the chain.")}
    ${V.moves === 0 && !V.last_move ? `<div class="sim-warn"><span class="sim-warn__icon">${icon("alert", 18)}</span><div><b>No move logged yet.</b> ${esc(V.history_note)}. Fire "A day of normal operation" on the Simulation tab and come back: the day's moves appear here with their stays.</div></div>` : ""}
    <div class="stats stats--5">
      ${stat(`Moves, last ${V.days} days`, "route", num(V.moves), V.moves ? `${num(V.pairs_count)} pairs of compartments` : esc(V.reason || ""), "", "stat__val--gold")}
      ${stat("Devices moved", "box", num(V.devices), V.devices ? `${num(Math.round(V.moves / V.devices * 10) / 10)} moves a device` : "")}
      ${stat("Busiest outflow", "layers", busiest && busiest.units_out ? esc(busiest.name) : "—", busiest && busiest.units_out ? `${num(busiest.units_out)} out · ${num(busiest.out_per_week)} a week` : "nothing left a compartment")}
      ${stat("First move logged", "clock", V.first_move ? fmtDate(V.first_move) : "—", V.last_move ? `last ${fmtDate(V.last_move)}` : "the log is empty")}
      ${stat("Without a day", "alert", num(V.undated_events), V.undated_events ? esc(V.undated_reason) : "every logged move carries its day", V.undated_events ? "stat__hint--neg" : "stat__hint--pos")}
    </div>
    <div class="section">
      <div class="section__head"><span class="section__title">Between compartments</span><span class="section__count">${V.pairs_count}</span>
        <div class="segmented" id="mv-window" style="margin-left:14px">${MV_WINDOWS.map((d) => `<button class="${d === MV.days ? "active" : ""}" data-days="${d}">${d} d</button>`).join("")}</div>
        <span class="section__hint" title="${esc(V.coverage_basis)}">since ${fmtDate(V.since)} · the log covers ${num(V.covered_days)} of the ${num(V.days)} days, a week's flow is over those · a stay is measured from the day a device entered the compartment to the day it left</span></div>
      <div class="panel" style="padding:${V.pairs.length ? 0 : 14}px 18px">${mvPairsHtml(V)}</div>
    </div>
    <div class="section">
      <div class="section__head"><span class="section__title">Per compartment</span><span class="section__count">${V.compartments.length}</span><span class="section__hint">measured from the moves, next to what the Warehouse tab derives from the stock still present</span></div>
      <div class="panel">${mvCompartmentsHtml(V)}</div>
    </div>
    <div class="section" style="margin-bottom:0">
      <div class="section__head"><span class="section__title">One device</span><span class="section__hint">a serial opens to its own path through the chain, with the days in each station</span></div>
      <div class="panel" style="padding:14px 18px">
        <form class="mv-form" id="mv-form"><input class="input" id="mv-serial" placeholder="Serial, e.g. DAAS-0000001" value="${esc(MV.serial)}" /><button type="submit" class="btn btn--ink btn--sm">${icon("route", 13)} Open</button></form>
        <div id="mv-path" style="margin-top:12px">${MV.path ? mvPathHtml(MV.path) : `<div class="muted" style="font-size:12.5px">Type a serial number.</div>`}</div>
      </div>
    </div>`;
  $$("#mv-window button").forEach((b) => b.addEventListener("click", () => { MV.days = Number(b.dataset.days); mvDraw(screen); }));
  $("#mv-form").addEventListener("submit", (e) => { e.preventDefault(); mvOpenSerial(); });
}

RENDER.movements = async function () {
  const screen = $("#screen");
  if (!isDaas()) {
    screen.innerHTML = `${pageHead("Warehouse", "Movements", "The movement log of the device fleet's warehouse.")}
      <div class="panel"><div class="state"><div class="state__icon">${icon("route", 22)}</div><div class="state__title">No compartments here</div>
      <div class="state__sub">This database holds the datacenter operation; the movement log follows the device fleet's compartments. Every asset's own event log is on the Assets tab.</div></div></div>`;
    return;
  }
  try {
    // A device can be linked to: /?serial=DAAS-0000001&days=90#movements opens its path.
    const q = new URLSearchParams(location.search);
    if (q.has("serial")) MV.serial = q.get("serial");
    if (MV_WINDOWS.includes(Number(q.get("days")))) MV.days = Number(q.get("days"));
    await mvDraw(screen);
    if (MV.serial && !MV.path) mvOpenSerial();
    // Live: a fleet event on the Simulation tab writes moves; one grouped read every 15 seconds while the tab is open.
    livePoll("movements", async () => {
      const V2 = await api(`/movements?days=${MV.days}`);
      if (MV.view && (V2.moves !== MV.view.moves || V2.devices !== MV.view.devices)) mvDraw(screen);
    }, 15000);
  } catch (e) { screen.innerHTML = errState(e.message); }
};
